src/OpenTelemetry/Resources/Resource.cs

```diff
@@ -102,61 +102,74 @@ public Resource Merge(Resource other)
                 sanitizedKey = attribute.Key;
             }
 
-            object sanitizedValue = SanitizeValue(attribute.Value, sanitizedKey);
+            var sanitizedValue = SanitizeValue(attribute.Value, sanitizedKey);
             return new KeyValuePair<string, object>(sanitizedKey, sanitizedValue);
         }
 
         private static object SanitizeValue(object value, string keyName)
         {
             Guard.Null(keyName, nameof(keyName));
 
-            if (value is string || value is bool || value is double || value is long)
+            return value switch
             {
-                return value;
-            }
-
-            if (value is string[] || value is bool[] || value is double[] || value is long[])
-            {
-                return value;
-            }
+                string => value,
+                bool => value,
+                double => value,
+                long => value,
+                string[] => value,
+                bool[] => value,
+                double[] => value,
+                long[] => value,
+                int => Convert.ToInt64(value),
+                short => Convert.ToInt64(value),
+                float => Convert.ToDouble(value, System.Globalization.CultureInfo.InvariantCulture),
+                int[] v => SanitizationUtils.ConvertoToInt64Array(v),
+                short[] v => SanitizationUtils.ConvertoToInt64Array(v),
+                float[] v => SanitizationUtils.ConvertToDoubleArray(v),
```

Comment:
Instead of these helper methods consider using the built-in [Array.ConvertAll ](https://docs.microsoft.com/en-us/dotnet/api/system.array.convertall?view=net-6.0) method

It might be useful to also check how the performance of this is compared to the current solution

Reply:
good point. thanks :)

from the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.array.convertall?view=net-6.0#remarks):

> This method is an O(n) operation, where n is the Length of array.

but I'll create a benchmark.net test checking allocations on these 3 options. (considering the code prior to this PR as the baseline)

Reply:
based on [this gist](https://gist.github.com/kiapanahi/1f74a39a6917474eb5a3e5e17a14bfa1) and the following results from a simple benchmark, I guess its safe to assume that `Array.ConvertAll` is pretty performant

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Xeon E-2176M CPU 2.70GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


```
|                                           Method |      Mean |     Error |    StdDev | Ratio |
|------------------------------------------------- |----------:|----------:|----------:|------:|
| &#39;Convert int[] to long[] using traditional loop&#39; | 54.787 ms | 1.0834 ms | 2.1886 ms |  1.00 |
| &#39;Convert int[] to long[] using Array.ConvertAll&#39; |  5.425 ms | 0.1179 ms | 0.3420 ms |  0.10 |

Reply:
Awesome! Looks great to me 😄 
One small nit comment is you could convert `ConvertToDoubleUsingIvariantCulture` into a lambda function since it is only used in that one place 👍

Reply:
You're absolutely right.
Although the static local function will be almost always aggressively inlined by JIT, the one-liner lambda notation is much easier on the eye.